### PR TITLE
[react] Allow more return values from stateless components

### PR DESCRIPTION
**Pull request by LinusU**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

-----

Opening this for some early feedback since I can't figure out how to tackle my last problem.

Basically, my original problem is that I have a stateless component that returns an array (it returns the return value of `React.Children.map`).

When fixing this I stumbled onto another problem; it's illegal to return `undefined` from a `render`-function, or from a stateless component. Yet this is currently allowed for normal components.

I think I got most of it right now, but for some reason I'm getting the following error:

```text
test/tsx.tsx(13,1): error TS2605: JSX element type 'ReactRenderResult' is not a constructor function for JSX elements.
  Type 'null' is not assignable to type 'ElementClass'.
```

whenever a stateless component is used in JSX:

```js
<StatelessComponent />;
```

Any help would be much appreciated! 🍻

### CODE REVIEW

File: types/react/test/index.ts [hunk 1]
```diff
@@ -708,3 +708,13 @@ class RenderChildren extends React.Component {
         return children !== undefined ? children : null;
     }
 }
+
+// Stateless functional components should be able to return bare arrays
+const ArrayElement: React.SFC = () => ([ 'Hello, ', 1337, '!']);
+const NullElement: React.SFC = () => null;
+
+// Render methods should not be allowed to return undefined
+// $ExpectError
```
Comment (vsiao) [inline types/react/test/index.ts]: I love that we can write these tests.

**Review by vsiao — APPROVED** [review]

**Comment by vsiao** [comment]
As for the SFC error, I think it may actually be a problem to be solved at the compiler level. There's some background in https://github.com/Microsoft/TypeScript/issues/5478 about how SFCs work. If @RyanCavanaugh could chime in here that would be best.

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.
